### PR TITLE
posix: add read timeout

### PR DESCRIPTION
**Pull request by liamstask**
working as expected for me on OS X and Linux. open to feedback, happy to make changes.

### CODE REVIEW

File: term_posix.go [hunk 1]
```diff
@@ -83,6 +84,42 @@ func (t *Term) setSpeed(baud int) error {
...
+	} else {
+		// block indefinitely until we receive at least 1 byte
+		vmin = 1
+	}
+	return
+}
+
+// SetReadTimeout sets the read timeout, after which io.EOF is returned.
+//
+// A zero value for d means read operations will not time out.
```
Comment (davecheney) [inline term_posix.go]: Can you please implement this following the pattern of SetSpeed above. That is a function ReadTimeout, that is applied with SetOption, and a helper, SetReadTimeout

**Comment by davecheney** [comment]
LGTM with the change suggested. Thanks

**Comment by liamstask** [comment]
added in db6b9aa

File: term_posix.go [hunk 2]
```diff
@@ -83,6 +84,53 @@ func (t *Term) setSpeed(baud int) error {
...
+		return lo
+	}
+	if v > hi {
+		return hi
+	}
+	return v
+}
+
+// timeoutVals converts d into values suitable for termios VMIN and VTIME ctrl chars
+func timeoutVals(d time.Duration) (vmin, vtime uint8) {
```
Comment (davecheney) [inline term_posix.go]: I would prefer if you didn't used named return values and naked returns

if d > 0 {
    // compute vtime
    return 0, vtime
} 
return 1, 0

File: term_posix.go [hunk 3]
```diff
@@ -83,6 +84,53 @@ func (t *Term) setSpeed(baud int) error {
...
+
+// ReadTimeout sets the read timeout option for the terminal.
+func ReadTimeout(d time.Duration) func(*Term) error {
+	return func(t *Term) error {
+		return t.setReadTimeout(d)
+	}
+}
+
+// SetReadTimeout sets the read timeout.
+//
```
Comment (davecheney) [inline term_posix.go]: please remove the blank line in the comment

**Comment by davecheney** [comment]
LGTM with minor nits

**Comment by liamstask** [comment]
added in 4b5cefa

let me know if you want me to squash any of these

File: term_posix.go [hunk 4]
```diff
@@ -83,6 +84,52 @@ func (t *Term) setSpeed(baud int) error {
...
+
+// timeoutVals converts d into values suitable for termios VMIN and VTIME ctrl chars
+func timeoutVals(d time.Duration) (uint8, uint8) {
+	if d > 0 {
+		// VTIME is expressed in terms of deciseconds
+		vtimeDeci := d.Nanoseconds() / 1e6 / 100
+		// ensure valid range
+		vtime := uint8(clamp(vtimeDeci, 1, 0xff))
+		return 0, vtime
+	} else {
```
Comment (davecheney) [inline term_posix.go]: Please drop the else block, the if block always ends in a return so else is not needed

**Comment by davecheney** [comment]
Really close, one last change. If you want to squash go for it.

**Comment by liamstask** [comment]
updated

**Comment by davecheney** [comment]
LGTM. Thank you very much.

**Comment by davecheney** [comment]
Btw. You may have seen the x/term repo has been created recently. It's likely that some of this package may be proposed as a strawman along with projects of other people's work.

**Comment by liamstask** [comment]
ah, great - i saw some of the discussion swirling around that. would be great to see a single well supported solution emerge, assuming everybody can agree on the api etc :)

**Comment by davecheney** [comment]
> assuming everybody can agree on the api etc

Ahh, so you've worked in open source before

**Pull request merged**